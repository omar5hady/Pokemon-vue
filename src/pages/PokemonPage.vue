<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1>Â¿Quien es este Pokemon?</h1>
        <div class="row">
            <h4 class="col-md-5">Puntuacion: {{puntuacion}}</h4>
            <h4 class="col-md-5">Vidas: 
                <img src="@/assets/vida.png" alt="" srcset="" width="20" height="20" :key="vida" v-for="vida in intentos">
            </h4>
        </div>
        <!-- Componente img Pokemon -->
        <PokemonImage :pokemonId="pokemon.id" :showPokemon="showPokemon"></PokemonImage>
        <!-- Componente opciones Pokemon -->
        <PokemonOpt 
            :pokemons="pokemonArr" 
            @selection-pokemon="checkAnswer($event)"
        ></PokemonOpt>

        <template v-if="showAnswer">
            <h2 class="fade-in">{{message}}</h2>
            <button @click="siguiente" v-if="intentos > 0">Otro pokemon</button>
            <button @click="newGame" v-if="intentos == 0">Nuevo juego</button>
        </template>
        
    </div>
</template>

<script>
import PokemonImage from '@/components/PokemonImage.vue'
import PokemonOpt from '@/components/PokemonOpt.vue'

import getPokemonOpt from '@/helpers/getPokemonsOpt'

//console.log( getPokemonOpt() )

export default {
    components:{
        PokemonImage,
        PokemonOpt,
    },
    data(){
        return {
            pokemonArr :[], //Arreglo para opciones pokemon
            pokemon : null,
            showPokemon : false,
            showAnswer : false,
            message : '',
            puntuacion : 0,
            intentos : 3,
            activo : true,
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = []
            this.pokemon = null
            this.pokemonArr = await getPokemonOpt()

            const rndInt = Math.floor ( Math.random() * 4)

            this.pokemon = this.pokemonArr[rndInt]
        },
        //$event es el valor que esta retornando el selection
        checkAnswer( pokemonSelected) {
            this.showPokemon = true
            this.showAnswer = true

            if(this.activo){
                if(this.pokemon.id === pokemonSelected){
                    this.message = `Correcto es ${this.pokemon.name}`
                    this.puntuacion ++
                }
                    
                else{
                    this.message = `No es ${this.pokemon.name}`
                    this.intentos --
                }
            }

            this.activo = false
            
            
        },
        newGame(){
            this.activo = true
            this.conteo = 0
            this.puntuacion = 0
            this.intentos = 3
            this.siguiente()
        },
        siguiente(){
            this.activo = true
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''
            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>
