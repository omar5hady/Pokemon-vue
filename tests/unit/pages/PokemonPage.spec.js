import { shallowMount, mount } from "@vue/test-utils";
import  PokemonPage  from "@/pages/PokemonPage";
import {pokemons} from '../mocks/pokemons.mock'

describe('PokemonPage Component', () =>{
    let wrapper 

    beforeEach( () => {
        wrapper = shallowMount(PokemonPage)
    })

    test('Debe hacer match con el snapshot', () => {
        expect( wrapper.html()).toMatchSnapshot()
    })

    test('Se llama metodo mixPokemonArray al montar el componente', () => {
        const mixPokemonArraySpay = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)
        expect( mixPokemonArraySpay ).toHaveBeenCalled()
    })

    test('Debe hacer match con el snapshot al cargar los pokemons', () =>{
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return{
                    pokemonArr :pokemons, //Arreglo para opciones pokemon
                    pokemon : pokemons[0],
                    showPokemon : false,
                    showAnswer : false,
                    message : '',
                    puntuacion : 0,
                    intentos : 3,
                    activo : true,
                }
            }
        })
        expect( wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar los componentes de PokemonImage y PokemonOpt', () =>{
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return{
                    pokemonArr :pokemons, //Arreglo para opciones pokemon
                    pokemon : pokemons[0],
                    showPokemon : false,
                    showAnswer : false,
                    message : '',
                    puntuacion : 0,
                    intentos : 3,
                    activo : true,
                }
            }
        })
        const pokemonImage = wrapper.findComponent('pokemon-image-stub')
        const pokemonOpt = wrapper.findComponent('pokemon-opt-stub')

        //PokemonPicture debe de existir
        expect(pokemonImage.exists()).toBe(true)
        //PokemonOpt debe de existir
        expect( pokemonOpt.exists()).toBe(true)

        //PokemonPicture attribute pokemonId === 1
        expect(pokemonImage.vm.pokemonId).toBe(1)

        //PokemonOpt attribute pokemons toBe true
        expect(pokemonOpt.vm.pokemons).toBeTruthy()
    })

    test('Pruebas con checkAnswer ',async ()=>{
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return{
                    pokemonArr :pokemons, //Arreglo para opciones pokemon
                    pokemon : pokemons[0],
                    showPokemon : false,
                    showAnswer : false,
                    message : '',
                    puntuacion : 0,
                    intentos : 3,
                    activo : false,
                }
            }
        })

        await wrapper.vm.checkAnswer(1)
        expect(wrapper.vm.showPokemon).toBeTruthy()

        console.log(wrapper.vm.pokemon)
    })
})