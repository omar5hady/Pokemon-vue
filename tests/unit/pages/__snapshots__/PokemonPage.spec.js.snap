// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage Component Debe hacer match con el snapshot al cargar los pokemons 1`] = `
<div>
  <h1>Â¿Quien es este Pokemon?</h1>
  <div class="row">
    <h4 class="col-md-5">Puntuacion: 0</h4>
    <h4 class="col-md-5">Vidas: <img src="" alt="" srcset="" width="20" height="20"><img src="" alt="" srcset="" width="20" height="20"><img src="" alt="" srcset="" width="20" height="20"></h4>
  </div><!-- Componente img Pokemon -->
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub><!-- Componente opciones Pokemon -->
  <pokemon-opt-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-opt-stub>
  <!--v-if-->
</div>
`;
